app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt

```diff
@@ -0,0 +1,240 @@
+		val intentBuilder = authService.createCustomTabsIntentBuilder(authRequest.get().toUri())
+		authIntent.set(intentBuilder.build())
+		authIntentLatch.countDown()
+	}
+
+	private fun createAuthRequest() {
+		Log.d(TAG, "Creating auth request")
+		val codeChallenge = getSpotifyPkceCodeChallenge(CODE_VERIFIER)
+		val codeChallengeMethod = "S256"
+		val clientId = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA)
+				.metaData.getString("com.spotify.music.API_KEY", "unavailable")
+		val authRequestBuilder = AuthorizationRequest.Builder(
+				authStateManager.getAuthorizationServiceConfiguration()!!,
+				clientId,
+				ResponseTypeValues.CODE,
+				Uri.parse(SpotifyAppController.REDIRECT_URI))
+				.setScopes(scopes)
+				.setCodeVerifier(CODE_VERIFIER, codeChallenge, codeChallengeMethod)
+		authRequest.set(authRequestBuilder.build())
+	}
+
+	private fun initializeAppAuth() {
+		Log.d(TAG, "Initializing AppAuth")
+		recreateAuthorizationService()
+
+		if (authStateManager.getAuthorizationServiceConfiguration() == null) {
+			val authEndpointUri = Uri.parse("https://accounts.spotify.com/authorize")
+			val tokenEndpointUri = Uri.parse("https://accounts.spotify.com/api/token")
+			val config = AuthorizationServiceConfiguration(authEndpointUri, tokenEndpointUri)
+			authStateManager.replaceAuthState(AuthState(config))
+		}
+
+		runBlocking {
+			createAuthRequest()
+			warmUpBrowser()
+		}
+	}
+
+	private fun recreateAuthorizationService() {
+		if (this::authService.isInitialized) {
+			Log.d(TAG, "Discarding existing AuthService instance")
+			authService.dispose()
+		}
+		authService = createAuthorizationService()
+		authRequest.set(null)
+		authIntent.set(null)
+	}
+
+	private fun createAuthorizationService(): AuthorizationService {
+		Log.d(TAG, "Creating authorization service")
+		val builder = AppAuthConfiguration.Builder()
+		builder.setBrowserMatcher(BrowserWhitelist(
+				VersionedBrowserMatcher.CHROME_CUSTOM_TAB,
+				VersionedBrowserMatcher.SAMSUNG_CUSTOM_TAB))
+		builder.setConnectionBuilder(DefaultConnectionBuilder.INSTANCE)
+		return AuthorizationService(this, builder.build())
+	}
+
+	private fun processAuthorizationCode(requestCode: Int, resultCode: Int, data: Intent?) {
+		if (requestCode == REQUEST_CODE_SPOTIFY_LOGIN) {
+			if (resultCode == RESULT_CANCELED) {
+				onAuthorizationCancelled()
+				return
+			}
+
+			data?.let { intent ->
+				val response = AuthorizationResponse.fromIntent(intent)
+				val ex = AuthorizationException.fromIntent(intent)
+				when {
+					response?.authorizationCode != null -> {
+						authStateManager.updateAuthorizationResponse(response, ex)
+						onAuthorizationSucceed()
+					}
+					ex != null -> {
+						onAuthorizationFailed("Authorization flow failed: " + ex.message)
+					}
+					else -> {
+						onAuthorizationFailed("No authorization state retained - reauthorization required")
+					}
+				}
+			}
+		}
+	}
+
+	override fun onCreate(savedInstanceState: Bundle?) {
```

Comment:
I like having these `override` methods at the top of the file, because they represent execution entrypoints and are handy to have at first glance at the top.

Reply:
That's a good point, moved override methods to top of file